### PR TITLE
Support `databuilder --version` on CLI

### LINKED ISSUE #624: Support `databuilder --version` on CLI
Unless anyone has better suggestions, this could be done in the same way as we current do for `opensafely-cli`.

1. Have a file at `databuilder/VERSION` whose default contents is `not-from-a-package` or `development` or something like that. (Bonus points if there's something approximating a convention we can follow here.)

2. Read that file at runtime and wire it up to the `--version` argument.

   https://github.com/opensafely-core/opensafely-cli/blob/a9efc8119a979a16fef1dd9ce1e470adad1be2c2/opensafely/__init__.py#L10
   ```py
   __version__ = Path(__file__).parent.joinpath("VERSION").read_text().strip()
   ```
   https://github.com/opensafely-core/opensafely-cli/blob/a9efc8119a979a16fef1dd9ce1e470adad1be2c2/opensafely/__init__.py#L21-L23
   ```py
       parser.add_argument(
           "--version", action="version", version=f"opensafely {__version__}"
       )
   ```

3. Have the action which builds the package write the correct version to the file before building.
   https://github.com/opensafely-core/opensafely-cli/blob/a9efc8119a979a16fef1dd9ce1e470adad1be2c2/.github/workflows/build_and_publish.yaml#L44-L46
   ```yaml
    - name: Generate correct value for VERSION file
      run: |
        echo ${{ needs.tag-new-version.outputs.tag }} > opensafely/VERSION
   ```

### LINKED ISSUE #731: Speed up Docker builds
#624 made Docker builds significantly slower, as Data Builder (and its dependencies) is installed on each build

https://github.com/opensafely-core/databuilder/blob/8fefe6c7578b92c75a17b27df3d561308ced1bea/Dockerfile#L33

to expose a version number. Given that builds are central to testing, this slowdown is intolerable. @evansd reported the issue in [this thread](https://bennettoxford.slack.com/archives/C03FB777L1M/p1662743210173519).

We should speed up Docker builds. We should also reconsider `setuptools_scm`.

### PR DESCRIPTION
**Pull request by iaindillingham**
This is the second attempt at adding the version argument. The first attempt (#676, #708), which was based on setuptools-scm, increased the time it took to build the Docker image. Here, we eschew setuptools-scm and simply write the version to a file. Consequently, the time it takes to build the Docker image remains the same.

Closes #624 
Closes #731

### CODE REVIEW

File: pyproject.toml [hunk 1]
```diff
@@ -1,11 +1,11 @@
...
-requires = ["setuptools>=45", "setuptools_scm[toml]>=6.2"]
-build-backend = "setuptools.build_meta"
+requires = ["flit_core >=3.2,<4"]
+build-backend = "flit_core.buildapi"
 
 [project]
 name = "opensafely-databuilder"
 description = ""
-dynamic = ["version"]
+version = "2+local"
```
Comment (ghickman) [inline pyproject.toml:8]: This will mean pip reports a different version to what's in the version file, was that intended?

**Review by ghickman — APPROVED** [review]

File: .github/workflows/build-and-deploy.yml [hunk 2]
```diff
@@ -98,7 +98,8 @@ jobs:
           echo "MINOR=$MINOR"
           echo "PATCH=$PATCH"
 
-          export DOCKER_BUILDKIT=1
```
Comment (StevenMaude) [inline .github/workflows/build-and-deploy.yml:101]: question: What's the reason for dropping BuildKit?

It was added to make the Dockerfile changes here work, I think. But it's the [more modern Docker way of building images](https://docs.docker.com/build/#overview), in any case. So maybe we still want to retain it?

We use it elsewhere; for example: https://github.com/opensafely-core/base-docker/blob/8d9de9f1c46816eca95b28f15e6aecc128f90b3e/Makefile#L4

**Comment by StevenMaude** [comment]
Because this reverts to flit for building, this might fix #724 whose cause might be a feature of/issue with setuptools_scm.

File: pyproject.toml (hunk 1 above)
Comment (iaindillingham) [reply pyproject.toml:8]: It wasn't intended. However, I'm unsure how to reconcile dynamic metadata with local version identifiers. Do you know, @ghickman?

File: .github/workflows/build-and-deploy.yml (hunk 2 above)
Comment (iaindillingham) [reply .github/workflows/build-and-deploy.yml:101]: The reason is just as you think: we weren't using it before, and there isn't an immediate need to use it now, so we drop it. We may choose to use it when implementing #665, though.

Comment (StevenMaude) [reply .github/workflows/build-and-deploy.yml:101]: :+1: — I think it's fine to leave this PR as is, then semantically re-introduce BuildKit as its own improvement.

File: pyproject.toml (hunk 1 above)
Comment (ghickman) [reply pyproject.toml:8]: @iaindillingham – you should be able to use your existing creation of a `__version__` attribute on the package to look it up [dynamically with flit](https://flit.pypa.io/en/latest/pyproject_toml.html#new-style-metadata).  Alternatively you could just use the same string for both 😁

**Pull request merged**